packages/js/product-editor/src/components/modal-editor/modal-editor.tsx

```diff
@@ -28,16 +29,40 @@ export function ModalEditor( {
 	onClose,
 	title,
 }: ModalEditorProps ) {
+	const [ isDebouncing, setIsDebouncing ] = useState( false );
+	const [ isClosing, setIsClosing ] = useState( false );
+
+	useEffect( () => {
+		// Only close the modal once editor changes have persisted after debouncing.
+		if ( isClosing && ! isDebouncing ) {
+			onClose();
+		}
+	}, [ isDebouncing, isClosing, onClose ] );
+
+	const debouncedOnChange = useDebounce( ( blocks: BlockInstance[] ) => {
+		onChange( blocks );
+		setIsDebouncing( false );
+	}, 250 );
+
+	function handleInput( blocks: BlockInstance[] ) {
+		setIsDebouncing( true );
+		debouncedOnChange( blocks );
+	}
+
+	function handleClose() {
+		setIsClosing( true );
+	}
+
 	return (
 		<Modal
 			className="woocommerce-modal-editor"
 			title={ title }
-			onRequestClose={ onClose }
+			onRequestClose={ handleClose }
 			shouldCloseOnClickOutside={ false }
 		>
 			<IframeEditor
 				initialBlocks={ initialBlocks }
-				onChange={ onChange }
+				onInput={ handleInput }
```

Comment:
It actually looks like you need both `onChange` and `onInput` :/ they are both being called for different circumstances.
For example, `onChange` only gets called when new blocks are created, but `onInput` when something within the block is updated. You can pretty easily see this if you console log both functions.
Also see this example: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/README.md?plain=1#L34-L35
And the description of the outputs: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/provider/README.md?plain=1#L14-L32

Reply:
Good call.  I consoled the outputs and noticed this as well.

>A callback invoked when the blocks have been modified in a non-persistent manner.

I had read the above and re-reading it I'm still convinced the documentation is semantically incorrect.  I would expect the non-persistent callback (`onInput`) to always fire while the persistent callback (`onChange`) would fire when those changes were persisted.

This behavior is all-around a little odd and does not match the [browser API](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event) as stated in the docs.  I'll open an issue upstream to get this fixed.